core/src/main/java/com/linecorp/armeria/server/ServiceNaming.java

```diff
@@ -68,7 +68,34 @@ static ServiceNaming fullTypeName() {
     static ServiceNaming simpleTypeName() {
         return ctx -> {
             final String fullTypeName = fullTypeName().serviceName(ctx);
-            return fullTypeName.substring(fullTypeName.lastIndexOf('.') + 1);
+            String simpleTypeName = "";
+            final int packageIndex = fullTypeName.lastIndexOf('.');
+            if (packageIndex >= 0) {
+                simpleTypeName = fullTypeName.substring(packageIndex + 1);
+            }
+
+            if (ctx.rpcRequest() != null) {
+                return simpleTypeName;
+            }
+
+            for (;;) {
```

Comment:
out of curiosity: IIUC, it seems to handle anonymous class, inner class or name-conflicted class such as `Foo$123`, `Foo$Bar` and `Foo$$`. In case of both anonymous and inner class, name after trailing dollars will be truncated. For anonymous class, it could be okay, but I am not sure that it is acceptable for inner class. For example, `Foo$Bar` become `Foo$` that cannot be distinguished, doesn't it?

Reply:
> In case of both anonymous and inner class, name after trailing dollars will be truncated.

Actually, this is the opposite case. Let's say we have the following nested class.

```java
package foo.bar;
class Outer {
    static class Inner {
        ...
    }
}
```
- `foo.bar.Outer.Inner.class.getName()` returns `"foo.bar.Outer$Inner"`
- `foo.bar.Outer.Inner.class.getSimpleName()` returns `"Inner"`

As per the Javadoc, I thought, we may need to return the same name pattern like `Class.getSimpleName()`. So if there is the outer class name in the full service name, it is truncated.
https://github.com/line/armeria/blob/2790d655f60c352ca47302c9b0d0fc8370e2d832/core/src/main/java/com/linecorp/armeria/server/ServiceNaming.java#L62-L67

Reply:
Sorry I was confused. Thanks for the explanation. Then, does it make sense for anonymous class? If anonymous class is defined with `FooService`, `FooService${numeric}` will be converted to `{numeric}`.

Reply:
> Currently, we return `"Outer$Inner"` for `ServiceNaming.simpleTypeName`.
> As per the Javadoc, I thought, we may need to return the same name pattern like `Class.getSimpleName()`. So if there is the outer class name in the full service name, it is truncated.

I think I prefer `Outer$Inner` and `Outer$1` vs. `Inner` and `1`. :smile:

Reply:
`1` sounds weird. 😆 Let me revert this logic and remove the `@see Class#getSimpleName()` also for less confusion. 
`It is supposed to have a class name without a period.` exactly explains how `simpleTypeName()` is made.